File: app/src/main/java/org/michaelbel/tjgram/ui/profile/ProfilePresenter.java
package org.michaelbel.tjgram.ui.profile;

import org.jetbrains.annotations.NotNull;
import org.michaelbel.tjgram.data.entity.User;
import org.michaelbel.tjgram.data.entity.UserResult;

import io.reactivex.disposables.CompositeDisposable;
import okhttp3.Headers;
import timber.log.Timber;

public class ProfilePresenter implements ProfileContract.Presenter {

    private ProfileContract.View view;
    private ProfileContract.Repository repository;
    private CompositeDisposable disposables = new CompositeDisposable();

    public ProfilePresenter(ProfileRepository repository) {
        this.repository = repository;
    }

    @NotNull
    @Override
    public ProfileContract.View getView() {
        return view;
    }

    @Override
    public void setView(@NotNull ProfileContract.View view) {
        this.view = view;
    }

    @Override
    public void authQr(@NotNull String token) {
        disposables.add(repository.authQr(token).subscribe(response -> {
            if (response.isSuccessful()) {
                Headers headers = response.headers();
                String xDeviceToken = headers.get("X-Device-Token");

                UserResult userResult = response.body();
                if (userResult != null) {
                    User user = userResult.getResult();
                    getView().setUser(user, xDeviceToken);
                }
            }
        }, throwable -> getView().setAuthError(throwable)));
    }

    @Override
    public void userMe() {
        disposables.add(repository.userMe().subscribe(userResult -> {
            User user = userResult.getResult();
            getView().setUser(user, "x");
        }, Timber::e));
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
    }
}

File: app/src/main/java/org/michaelbel/tjgram/modules/profile/ProfilePresenter.kt
package org.michaelbel.tjgram.modules.profile

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import org.michaelbel.tjgram.data.remote.TjService
import timber.log.Timber

class ProfilePresenter(private val service: TjService) : ProfileContract.Presenter {

    private var view: ProfileContract.View? = null
    private val disposables = CompositeDisposable()

    override fun create(view: ProfileContract.View) {
        this.view = view
    }

    override fun authQr(token: String) {
        disposables.add(service.authQr(token).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({ response ->
            if (response.isSuccessful) {
                val headers = response.headers()
                val xDeviceToken = headers.get("X-Device-Token")

                val userResult = response.body()
                if (userResult != null) {
                    val user = userResult.result
                    view!!.setUserMe(user!!, xDeviceToken)
                }
            }
        }, { throwable -> view!!.setAuthError(throwable) }))
    }

    override fun userMe() {
        disposables.add(service.userMe().subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread()).subscribe({ userResult ->
            val user = userResult.result
            view!!.setUserMe(user!!, "x")
        },  { Timber.e(it) }))
    }

    override fun destroy() {
        disposables.dispose()
    }
}